### PR TITLE
Circuit with key rotation pytests

### PR DESCRIPTION
**Pull request by sgonta**

### CODE REVIEW

File: qa/test_framework/util.py [hunk 1]
```diff
@@ -494,19 +495,31 @@ def get_coinbase_address(node, expected_utxos=None):
...
-    fe_certificate_field_configs = certificate_field_config_csw_disabled
-    if is_csw_enabled:
-        fe_certificate_field_configs = certificate_field_config_csw_enabled
+    if type_of_circuit_number == 0:
+        fe_certificate_field_configs = certificate_field_config_csw_disabled
+
+        if is_csw_enabled:
+            fe_certificate_field_configs = certificate_field_config_csw_enabled
+    else:
+        fe_certificate_field_configs = certificate_with_key_rotation_field_config
```
Comment (i-Alex) [inline qa/test_framework/util.py:522]: I propose to add more comments, explaining this tricky configs.
And moreover, we may restructure a bit:
```
if type_of_circuit_number == 0:
  assert sc_creation_version <= 1
  fe_certificate_field_configs = ce...

elif type_of_circuit_number == 1:
  assert sc_creation_version == 2
  assert is_csw_enabled is False
  fe_certificate_field_configs = certificate_with_key_rotation_field_config
  
```

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 2]
```diff
@@ -343,6 +346,130 @@ private void processGenerateCertProofInfo(JsonNode json) {
...
+
+            signersPublicKeys.add(pkNode.asText());
+        }
+
+        if (!json.has("mastersPublicKeys") || !json.get("mastersPublicKeys").isArray()) {
+            printGenerateCertProofInfoUsageMsg("wrong public keys");
+            return;
+        }
+
+        List<String> mastersPublicKeys = new ArrayList<String>();
```
Comment (i-Alex) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:374]: check that `mastersPublicKeys` size equals to `signersPublicKeys` and throw proper error.

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 3]
```diff
@@ -343,6 +346,130 @@ private void processGenerateCertProofInfo(JsonNode json) {
...
+            printGenerateCertProofInfoUsageMsg("wrong verificationKeyPath value. Textual value expected.");
+            return;
+        }
+        String verificationKeyPath = json.get("verificationKeyPath").asText();
+
+        if (!json.has("isCSWEnabled") || !json.get("isCSWEnabled").isBoolean()) {
+            printGenerateCertProofInfoUsageMsg("wrong isCSWEnabled value. Boolean value expected.");
+            return;
+        }
+        boolean isCSWEnabled = json.get("isCSWEnabled").asBoolean();
```
Comment (i-Alex) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:416]: can be removed. Not a case for new circuit.

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 4]
```diff
@@ -343,6 +346,130 @@ private void processGenerateCertProofInfo(JsonNode json) {
...
+
+            if (!initDlogKey()) {
+                printer.print("Error occurred during dlog key generation.");
+                return;
+            }
+
+            int numOfCustomFields = CommonCircuit.CUSTOM_FIELDS_NUMBER_WITH_DISABLED_CSW;
+            if (isCSWEnabled){
+                numOfCustomFields = CommonCircuit.CUSTOM_FIELDS_NUMBER_WITH_ENABLED_CSW;
+            }
```
Comment (i-Alex) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:432]: For new circuit we have another constant with value `1`

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 5]
```diff
@@ -343,6 +346,130 @@ private void processGenerateCertProofInfo(JsonNode json) {
...
+        resJson.put("verificationKey", verificationKey);
+
+        ArrayNode keyArrayNode = resJson.putArray("schnorrPublicKeys");
+
+        for (String publicKeyStr : signersPublicKeys) {
+            keyArrayNode.add(publicKeyStr);
+        }
+        for (String publicKeyStr : mastersPublicKeys) {
+            keyArrayNode.add(publicKeyStr);
+        }
```
Comment (i-Alex) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:467]: separate them into signersPublicKeys and mastersPublicKeys.
In STF they must be used in different way

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: qa/SidechainTestFramework/scutil.py [hunk 6]
```diff
@@ -267,31 +267,36 @@ def generate_cert_signer_secrets(seed, number_of_schnorr_keys):
...
     signer_secrets = []
-    signer_public_keys = []
+    public_signing_keys = []
+    public_master_keys = []
     for i in range(len(signer_keys)):
         keys = signer_keys[i]
         signer_secrets.append(keys.secret)
-        signer_public_keys.append(keys.publicKey)
+        public_signing_keys.append(keys.publicKey)
+        public_master_keys.append(keys.publicKey)
```
Comment (MarcoOl94) [inline qa/SidechainTestFramework/scutil.py:281]: Master public keys should be generated in the same way of the signer keys, they should not be the same.

File: qa/resources/template_predefined_genesis.conf [hunk 7]
```diff
@@ -46,7 +46,7 @@ sparkz {
     signersThreshold = 5
```
Comment (MarcoOl94) [inline qa/resources/template_predefined_genesis.conf:46]: Add also here the master public keys placeholder

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

**Comment by sgonta** [comment]
Changes are merged into mo/key_rotation branch

**Pull request closed without merging**